application/src/ext/java/org/opentripplanner/ext/flex/AreaStopsToVerticesMapper.java

```diff
@@ -46,50 +52,65 @@ public void buildGraph() {
     );
 
     LOG.info(progress.startMessage());
-    var results = timetableRepository
+
+    timetableRepository
       .getSiteRepository()
       .listAreaStops()
       .parallelStream()
-      .flatMap(areaStop -> {
-        var matchedVertices = matchingVerticesForStop(graph, areaStop);
+      .forEach(areaStop -> {
+        applyAreaStopToVertices(areaStop, graph);
         // Keep lambda! A method-ref would cause incorrect class and line number to be logged
         progress.step(m -> LOG.info(m));
-        return matchedVertices;
-      });
-
-    ImmutableMultimap<StreetVertex, AreaStop> mappedResults = results.collect(
-      ImmutableListMultimap.<MatchResult, StreetVertex, AreaStop>flatteningToImmutableListMultimap(
-        MatchResult::vertex,
-        mr -> Stream.of(mr.stop())
-      )
-    );
-
-    mappedResults
-      .keySet()
-      .forEach(vertex -> {
-        vertex.addAreaStops(mappedResults.get(vertex));
       });
 
     LOG.info(progress.completeMessage());
   }
 
-  private static Stream<MatchResult> matchingVerticesForStop(Graph graph, AreaStop areaStop) {
-    return graph
-      .findVertices(areaStop.getGeometry().getEnvelopeInternal())
-      .stream()
+  private static void applyAreaStopToVertices(AreaStop areaStop, Graph graph) {
+    var vertices = graph.findVertices(areaStop.getGeometry().getEnvelopeInternal());
+
+    @Nullable
+    var progress = progressTracker(areaStop, vertices);
+
+    var geometryFactory = GeometryUtils.getGeometryFactory();
+
+    vertices
+      .parallelStream()
       .filter(StreetVertex.class::isInstance)
       .map(StreetVertex.class::cast)
       .filter(StreetVertex::isEligibleForCarPickupDropoff)
-      .filter(vertx -> {
+      .forEach(vertx -> {
         // The street index overselects, so need to check for exact geometry inclusion
-        Point p = GeometryUtils.getGeometryFactory().createPoint(vertx.getCoordinate());
-        return areaStop.getGeometry().intersects(p);
-      })
-      .map(vertx -> new MatchResult(vertx, areaStop));
+        Point p = geometryFactory.createPoint(vertx.getCoordinate());
+        var intersects = areaStop.getGeometry().intersects(p);
+        if (progress != null) {
+          // Keep lambda! A method-ref would cause incorrect class and line number to be logged
+          progress.step(m -> LOG.info(m));
+        }
+        if (intersects) {
+          vertx.addAreaStops(List.of(areaStop));
+        }
+      });
+    if (progress != null) {
+      LOG.info(progress.completeMessage());
+    }
   }
 
   /**
-   * The result of an area stop being matched with a vertex.
+   * We want a vertex progress tracker only for large areas, otherwise it becomes too noisy.
```

Comment:
Consider using one progress tracker, not one for each large areas. The tracker is thread safe, pass the tracker into  `applyAreaStopToVertices()`.